File: attendx/frontend/lib/services/attendance_engine.dart
import 'dart:math';
import 'local_db.dart';
import '../models/subject.dart';
import '../models/timetable.dart';
import '../models/attendance.dart';
import '../models/dashboard.dart';

class AttendanceEngine {
  final LocalDb db = LocalDb.instance;

  Future<Map<String, dynamic>?> getActiveTimetableVersion(DateTime date) async {
    final versions = await db.getTimetableVersions();
    final dStr = date.toIso8601String().split('T').first;
    for (var v in versions) {
      if ((v['effective_from'] as String).compareTo(dStr) <= 0) {
        return v;
      }
    }
    return null;
  }

  Future<bool> isHoliday(DateTime date) async {
    final weekday = date.weekday; // 1=Mon, 7=Sun
    if (weekday == 7) return true;
    if (weekday == 6) {
      final sem = await db.getSemester();
      if (sem != null && sem['saturday_enabled'] == 1) return false;
      return true;
    }
    return false;
  }

  Future<List<Map<String, dynamic>>> getSlotsForDate(DateTime date) async {
    if (await isHoliday(date)) return [];
    final v = await getActiveTimetableVersion(date);
    if (v == null) return [];
    
    // In Dart DateTime, weekday is 1=Mon..7=Sun
    // In DB, day_of_week is 0=Mon..6=Sun
    final dbWeekday = date.weekday - 1;
    
    final database = await db.database;
    final slots = await database.query(
      'class_slots',
      where: 'timetable_version_id = ? AND day_of_week = ?',
      whereArgs: [v['id'], dbWeekday],
    );
    return slots;
  }

  Future<List<Map<String, dynamic>>> ensureAttendanceForDate(DateTime date) async {
    final dStr = date.toIso8601String().split('T').first;
    final slots = await getSlotsForDate(date);
    final created = <Map<String, dynamic>>[];
    final database = await db.database;

    for (var slot in slots) {
      final existing = await database.query(
        'attendance_records',
        where: 'class_slot_id = ? AND date = ?',
        whereArgs: [slot['id'], dStr],
      );
      
      if (existing.isEmpty) {
        final rec = {
          'class_slot_id': slot['id'],
          'date': dStr,
          'status': 'PRESENT',
          'units': slot['duration_hours'],
          'auto_marked': 1,
        };
        final id = await database.insert('attendance_records', rec);
        rec['id'] = id;
        created.add(rec);
      }
    }
    return created;
  }

  Future<void> autoMarkRange(DateTime start, DateTime end) async {
    var d = start;
    while (d.compareTo(end) <= 0) {
      await ensureAttendanceForDate(d);
      d = d.add(const Duration(days: 1));
    }
  }

  Future<Map<String, dynamic>> computeSubjectStats(Subject subject, {DateTime? upTo}) async {
    final database = await db.database;
    var query = '''
      SELECT a.* FROM attendance_records a
      JOIN class_slots c ON a.class_slot_id = c.id
      WHERE c.subject_id = ?
    ''';
    var args = <dynamic>[subject.id];

    if (upTo != null) {
      query += ' AND a.date <= ?';
      args.add(upTo.toIso8601String().split('T').first);
    }

    final records = await database.rawQuery(query, args);

    double totalClasses = 0;
    double attended = 0;

    for (var r in records) {
      final status = r['status'] as String;
      final units = (r['units'] as num).toDouble();
      if (status != 'NO_CLASS') {
        totalClasses += units;
        if (status == 'PRESENT') attended += units;
      }
    }

    final percentage = totalClasses > 0 ? (attended / totalClasses * 100) : 0.0;
    final reqFrac = subject.requiredPercentage / 100.0;
    
    int safeBunks = 0;
    if (reqFrac > 0) {
      safeBunks = ((attended / reqFrac) - totalClasses).floor();
    }
    safeBunks = percentage >= subject.requiredPercentage ? max(safeBunks, 0) : min(safeBunks, 0);

    String statusStr = 'CRITICAL';
    if (percentage >= subject.requiredPercentage) {
      statusStr = 'SAFE';
    } else if (percentage >= subject.requiredPercentage - 10) {
      statusStr = 'WARNING';
    }

    return {
      'total_classes': totalClasses,
      'attended_classes': attended,
      'percentage': percentage,
      'safe_bunks': safeBunks,
      'status': statusStr,
    };
  }

  int? classesNeededToRecover(double attended, double total, double reqFrac) {
    if (reqFrac <= 0 || reqFrac >= 1) return null;
    final current = total > 0 ? attended / total : 0.0;
    if (current >= reqFrac) return null;
    
    final num = reqFrac * total - attended;
    final den = 1 - reqFrac;
    final x = num / den;
    return max(x.ceil(), 0);
  }

  Future<DashboardSummary> getDashboardSummary() async {
    final subjectsData = await db.getSubjects();
    final List<SubjectDashboard> rows = [];
    double totalAttended = 0;
    double totalClasses = 0;

    for (var sMap in subjectsData) {
      final subject = Subject.fromJson(sMap);
      final stats = await computeSubjectStats(subject);
      rows.add(SubjectDashboard(
        subjectId: subject.id,
        name: subject.name,
        code: subject.code,
        color: subject.color,
        requiredPercentage: subject.requiredPercentage,
        totalClasses: stats['total_classes'],
        attendedClasses: stats['attended_classes'],
        percentage: stats['percentage'],
        safeBunks: stats['safe_bunks'],
        status: stats['status'],
      ));
      totalAttended += stats['attended_classes'];
      totalClasses += stats['total_classes'];
    }

    final overall = totalClasses > 0 ? (totalAttended / totalClasses * 100) : 0.0;
    return DashboardSummary(
      overallPercentage: overall,
      subjects: rows,
    );
  }

  Future<List<Map<String, dynamic>>> getCalendarMonth(int year, int month) async {
    final daysInMonth = DateTime(year, month + 1, 0).day;
    final result = <Map<String, dynamic>>[];
    final database = await db.database;

    for (var day = 1; day <= daysInMonth; day++) {
      final d = DateTime(year, month, day);
      final dStr = d.toIso8601String().split('T').first;
      final slots = await getSlotsForDate(d);
      
      if (slots.isEmpty) {
        result.add({'date': dStr, 'summary': 'HOLIDAY'});
        continue;
      }

      final records = await database.rawQuery('''
        SELECT a.status FROM attendance_records a
        JOIN class_slots c ON a.class_slot_id = c.id
        WHERE a.date = ?
      ''', [dStr]);

      final statuses = records.map((r) => r['status'] as String).toSet();
      String summary;
      if (statuses.isEmpty) {
        summary = 'PENDING';
      } else if (statuses.length == 1 && statuses.contains('PRESENT')) {
        summary = 'ALL_PRESENT';
      } else if (statuses.contains('ABSENT')) {
        summary = 'HAS_ABSENT';
      } else {
        summary = 'MIXED';
      }
      result.add({'date': dStr, 'summary': summary});
    }
    return result;
  }

  Future<List<AttendanceRecord>> getDayAttendance(DateTime date) async {
    await ensureAttendanceForDate(date);
    final dStr = date.toIso8601String().split('T').first;
    final database = await db.database;
    
    final records = await database.rawQuery('''
      SELECT a.*, c.subject_id, s.name as subject_name, c.start_time, c.end_time 
      FROM attendance_records a
      JOIN class_slots c ON a.class_slot_id = c.id
      JOIN subjects s ON c.subject_id = s.id
      WHERE a.date = ?
    ''', [dStr]);
    
    return records.map((r) => AttendanceRecord.fromJson(r)).toList();
  }

  Future<void> overrideAttendance(int recordId, AttendanceStatus status) async {
    final database = await db.database;
    await database.update(
      'attendance_records',
      {'status': statusToString(status), 'auto_marked': 0},
      where: 'id = ?',
      whereArgs: [recordId],
    );
  }

  Future<Map<String, dynamic>> getTrend(int subjectId) async {
    final database = await db.database;
    final subjectMap = (await database.query('subjects', where: 'id = ?', whereArgs: [subjectId])).first;
    final subject = Subject.fromJson(subjectMap);

    final records = await database.rawQuery('''
      SELECT a.* FROM attendance_records a
      JOIN class_slots c ON a.class_slot_id = c.id
      WHERE c.subject_id = ?
      ORDER BY a.date ASC
    ''', [subjectId]);

    final trend = <Map<String, dynamic>>[];
    double cumTotal = 0;
    double cumAttended = 0;

    for (var r in records) {
      final status = r['status'] as String;
      if (status == 'NO_CLASS') continue;
      
      final units = (r['units'] as num).toDouble();
      cumTotal += units;
      if (status == 'PRESENT') cumAttended += units;
      
      final pct = cumTotal > 0 ? (cumAttended / cumTotal * 100) : 0.0;
      trend.add({'date': r['date'], 'cumulative_percentage': pct});
    }

    final stats = await computeSubjectStats(subject);
    final reqFrac = subject.requiredPercentage / 100.0;
    final below = stats['percentage'] < subject.requiredPercentage;
    final recover = below ? classesNeededToRecover(stats['attended_classes'], stats['total_classes'], reqFrac) : null;

    return {
      'subject_id': subject.id,
      'trend': trend,
      'is_below_required': below,
      'classes_needed_to_recover': recover,
      'max_missable_now': stats['safe_bunks'],
    };
  }

  Future<Map<String, dynamic>> predict(int subjectId, {required int missed}) async {
    final database = await db.database;
    final subjectMap = (await database.query('subjects', where: 'id = ?', whereArgs: [subjectId])).first;
    final subject = Subject.fromJson(subjectMap);
    
    final stats = await computeSubjectStats(subject);
    
    final projectedTotal = stats['total_classes'] + missed;
    final projectedAttended = stats['attended_classes']; // Missed all future
    
    final projectedPct = projectedTotal > 0 ? (projectedAttended / projectedTotal * 100) : 0.0;
    
    return {
      'subject_id': subject.id,
      'current_percentage': stats['percentage'],
      'projected_percentage': projectedPct,
      'required_percentage': subject.requiredPercentage,
      'will_be_below_required': projectedPct < subject.requiredPercentage,
    };
  }
}

File: attendx/backend/app/services/attendance_engine.py
"""
Core attendance logic shared by routers:
- resolving which TimetableVersion applies on a given date
- auto-marking PRESENT for past/today's classes
- computing dashboard stats & the safe-bunk formula
- analytics / trend / prediction
"""
import math
from datetime import date, timedelta
from typing import Optional, List

from sqlalchemy.orm import Session
from sqlalchemy import asc

from .. import models


def get_active_timetable_version(db: Session, on_date: date) -> Optional[models.TimetableVersion]:
    """The active version is the most recent one with effective_from <= on_date."""
    return (
        db.query(models.TimetableVersion)
        .filter(models.TimetableVersion.effective_from <= on_date)
        .order_by(models.TimetableVersion.effective_from.desc())
        .first()
    )


def is_holiday(db: Session, on_date: date) -> bool:
    """Sunday always a holiday. Saturday holiday unless enabled on the semester."""
    weekday = on_date.weekday()  # Mon=0..Sun=6
    if weekday == 6:
        return True
    if weekday == 5:
        semester = db.query(models.Semester).filter(models.Semester.is_active == True).first()  # noqa: E712
        saturday_enabled = semester.saturday_enabled if semester else False
        return not saturday_enabled
    return False


def get_slots_for_date(db: Session, on_date: date) -> List[models.ClassSlot]:
    if is_holiday(db, on_date):
        return []
    version = get_active_timetable_version(db, on_date)
    if not version:
        return []
    weekday = on_date.weekday()
    return [s for s in version.class_slots if s.day_of_week == weekday]


def ensure_attendance_records_for_date(db: Session, on_date: date) -> List[models.AttendanceRecord]:
    """
    Auto-mark step. For every class slot scheduled on `on_date` that does not
    already have an AttendanceRecord, create one with status=PRESENT (auto_marked=True).
    Existing records (manual or previously auto-marked) are left untouched.
    """
    slots = get_slots_for_date(db, on_date)
    created = []
    for slot in slots:
        existing = (
            db.query(models.AttendanceRecord)
            .filter(models.AttendanceRecord.class_slot_id == slot.id, models.AttendanceRecord.date == on_date)
            .first()
        )
        if existing:
            continue
        record = models.AttendanceRecord(
            class_slot_id=slot.id,
            date=on_date,
            status=models.AttendanceStatus.PRESENT,
            units=slot.duration_hours,
            auto_marked=True,
        )
        db.add(record)
        created.append(record)
    if created:
        db.commit()
    return created


def auto_mark_range(db: Session, start: date, end: date):
    """Run ensure_attendance_records_for_date for every day in [start, end]."""
    d = start
    total_created = 0
    while d <= end:
        total_created += len(ensure_attendance_records_for_date(db, d))
        d += timedelta(days=1)
    return total_created


def compute_subject_stats(db: Session, subject: models.Subject, upto_date: Optional[date] = None):
    """
    total_classes  = count of records with status != NO_CLASS
    attended       = count of records with status == PRESENT (in attendance UNITS, i.e. duration weighted)
    percentage     = attended / total * 100
    safe_bunks     = floor(attended / required_fraction - total_classes)
                     (required_fraction = required_percentage / 100)
    """
    q = (
        db.query(models.AttendanceRecord)
        .join(models.ClassSlot, models.AttendanceRecord.class_slot_id == models.ClassSlot.id)
        .filter(models.ClassSlot.subject_id == subject.id)
    )
    if upto_date:
        q = q.filter(models.AttendanceRecord.date <= upto_date)
    records = q.all()

    total_classes = sum(r.units for r in records if r.status != models.AttendanceStatus.NO_CLASS)
    attended = sum(r.units for r in records if r.status == models.AttendanceStatus.PRESENT)

    percentage = (attended / total_classes * 100) if total_classes > 0 else 0.0
    required_fraction = (subject.required_percentage or 75.0) / 100.0

    if required_fraction > 0:
        safe_bunks = math.floor(attended / required_fraction - total_classes)
    else:
        safe_bunks = 0
    safe_bunks = max(safe_bunks, 0) if percentage >= subject.required_percentage else min(safe_bunks, 0)

    if percentage >= subject.required_percentage:
        status = "SAFE"
    elif percentage >= subject.required_percentage - 10:
        status = "WARNING"
    else:
        status = "CRITICAL"

    return {
        "total_classes": total_classes,
        "attended_classes": attended,
        "percentage": round(percentage, 2),
        "safe_bunks": safe_bunks,
        "status": status,
    }


def classes_needed_to_recover(attended: float, total: float, required_fraction: float) -> Optional[int]:
    """
    How many additional classes (assuming all attended) are needed for the
    percentage to reach required_fraction again. Returns None if already met.
    Derivation: (attended + x) / (total + x) >= required_fraction
                => x >= (required_fraction*total - attended) / (1 - required_fraction)
    """
    if required_fraction <= 0 or required_fraction >= 1:
        return None
    current = attended / total if total > 0 else 0
    if current >= required_fraction:
        return None
    numerator = required_fraction * total - attended
    denominator = 1 - required_fraction
    x = numerator / denominator
    return max(math.ceil(x), 0)


def build_trend(db: Session, subject: models.Subject) -> List[dict]:
    q = (
        db.query(models.AttendanceRecord)
        .join(models.ClassSlot, models.AttendanceRecord.class_slot_id == models.ClassSlot.id)
        .filter(models.ClassSlot.subject_id == subject.id)
        .order_by(asc(models.AttendanceRecord.date))
    )
    records = q.all()
    trend = []
    cum_total = 0.0
    cum_attended = 0.0
    for r in records:
        if r.status == models.AttendanceStatus.NO_CLASS:
            continue
        cum_total += r.units
        if r.status == models.AttendanceStatus.PRESENT:
            cum_attended += r.units
        pct = (cum_attended / cum_total * 100) if cum_total > 0 else 0.0
        trend.append({"date": r.date, "cumulative_percentage": round(pct, 2)})
    return trend
